package/src/error.rs

```diff
@@ -0,0 +1,188 @@
+    },
+    NoPackageRoot {
+        path: PathBuf,
+    },
+    NoProjectDir,
+    RestrictedPath {
+        /// The url of the git package that tried the bad import.
+        package_url: Box<gix::Url>,
+        /// The git id of the bad package.
+        package_commit: ObjectId,
+        /// The relative path of the bad package within its git repo.
+        package_path: PathBuf,
+        attempted: PathBuf,
+        restriction: PathBuf,
+    },
+    /// There was some error interacting with a git repository.
+    Git(nickel_lang_git::Error),
+    InvalidUrl {
+        url: String,
+        msg: String,
+    },
+    Resolution {
+        msg: String,
+    },
+    InternalManifestError {
+        path: PathBuf,
+        msg: String,
+    },
+    /// There was an error persisting a temporary file.
+    TempFilePersist {
+        error: tempfile::PersistError,
+    },
+    NoManifestParent,
+    /// Index dependencies aren't implemented yet, so we emit
+    /// this if we encounter one.
+    IndexDep,
+}
+
+impl std::error::Error for Error {}
+
+impl std::fmt::Debug for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        std::fmt::Display::fmt(self, f)
+    }
+}
+
+impl std::fmt::Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Error::Io { error, path } => {
+                if let Some(path) = path {
+                    write!(f, "{}: {error}", path.display())
+                } else {
+                    error.fmt(f)
+                }
+            }
+            // Just a short and not-very-informative error. To write a better error message to
+            // the terminal, use `program.report` like the cli does.
+            Error::ManifestEval { package, .. } => {
+                if let Some(package) = package {
+                    write!(f, "error evaluating manifest for package {package}")
+                } else {
+                    write!(f, "error evaluating package manifest")
+                }
+            }
+            Error::RestrictedPath {
+                attempted,
+                restriction,
+                package_url,
+                package_commit,
+                package_path,
+            } => {
+                write!(
+                    f,
+                    "git package {package_url}@{package_commit}/{} tried to import path {}, but can only import from {}",
+                    package_path.display(),
+                    attempted.display(),
+                    restriction.display()
+                )
+            }
+            Error::Git(e) => e.fmt(f),
+            Error::InvalidUrl { msg, .. } => {
+                write!(f, "{}", msg)
+            }
+            Error::InternalManifestError { path, msg } => {
+                write!(
+                    f,
+                    "internal error reading the manifest at {}; this is a bug in nickel: {msg}",
```

Comment:
TODO: reuse the same internal error message that is used in Nickel-lang core. I think there is also a note saying to report the error to the issue tracker.